### ISSUE TITLE
False positive in UnnecessaryApply

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#2938]
## Expected Behavior
If we use the return value of `apply`, the `UnnecessaryApply` rule should never flag that usage.

## Observed Behavior
`UnnecessaryApply` flags some of those.

## Steps to Reproduce
```kotlin
val a = listOf(mutableListOf(""))
    .map { it.apply { add("") } }
```

## Context
Reported by zmunm in slack: https://kotlinlang.slack.com/archives/C88E12QH4/p1596599789239000

I really don't like this code because it goes agains immutability. But it's legit code so we should not flag it in this rule.

## Your Environment
* Version of detekt used: master, 1.11.0-RC1

**Comment by zsmoore** [comment]
I'll jump into this one that we are also seeing this when doing a 

object = data?.let { data -> new Object().apply { setData(data) } }
swapped to also temporarily to avoid checkstyle error

**Comment by BraisGabin** [comment]
Yes, I think this is the same case. You are using the return of `apply` as the return of `let`. And `UnnecessaryApply` doesn't check that.

**Comment by cortinico** [comment]
> If we use the return value of `apply`, the `UnnecessaryApply` rule should never flag that usage.

Turns out we already have the logic to do this, but only if type resolution is active. I'm restricting `UnnecessaryApply` to run only when type resolution is active to avoid this false positive here #2963
